esp8266/os_timer.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015 Rob Fowler
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stddef.h>
+
+#include "py/nlr.h"
+#include "py/runtime.h"
+#include "py/gc.h"
+
+#include "osapi.h"
+#include "os_type.h"
+#include "gccollect.h"
+#include "utils.h"
+
+#include "os_timer.h"
+
+extern void ets_timer_disarm(ETSTimer *ptimer);
+extern void ets_timer_setfn(ETSTimer *ptimer, ETSTimerFunc *pfunction, void *parg);
+extern void ets_timer_arm_new(ETSTimer *ptimer,uint32_t milliseconds, bool repeat_flag, bool);
+
+typedef struct _esp_os_timer_obj_t {
+    mp_obj_base_t base;
+    os_timer_t timer;
+    mp_obj_t callback;
+    uint32_t period;    // in milliseconds
+    bool repeat;
+    // start_time; // to calculate elapsed, time to fire etc
+} esp_os_timer_obj_t;
+
+
+STATIC void esp_os_timer_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    esp_os_timer_obj_t *self = self_in;
+
+    mp_printf(print, "timer(callback=%p, period=%u, repeat=%c)", 
+        (unsigned int)&self->callback,
+        self->period,
+        self->repeat ? 'y' : 'n');
+}
+
+STATIC void timer_common_callback(void *parg) {
+    esp_os_timer_obj_t *self = (esp_os_timer_obj_t *)parg;
+    if (self->callback) {
+        nlr_buf_t nlr;
+        if (nlr_push(&nlr) == 0) {
+            (void)mp_call_function_1(self->callback, self);
+        } else {
+            mp_obj_print_exception(&mp_plat_print, (mp_obj_t)nlr.ret_val);
+        }
+        gc_collect();       // done out of timed job and are going back to the operating command and into non determinate timing
```

Comment:
As this is in the toy operating system thread and the repl that is part of the console is never called. I can't see anywhere the GC will be called otherwise.